compiler/rustc_borrowck/src/type_check/mod.rs

```diff
@@ -2319,7 +2319,41 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
                         let cast_ty_from = CastTy::from_ty(ty_from);
                         let cast_ty_to = CastTy::from_ty(*ty);
                         match (cast_ty_from, cast_ty_to) {
-                            (Some(CastTy::Ptr(_)), Some(CastTy::Ptr(_))) => (),
+                            (Some(CastTy::Ptr(src)), Some(CastTy::Ptr(dst))) => {
+                                let src_tail = tcx.struct_tail_without_normalization(src.ty);
+                                let dst_tail = tcx.struct_tail_without_normalization(dst.ty);
+
+                                if let ty::Dynamic(..) = src_tail.kind()
+                                    && let ty::Dynamic(dst_tty, ..) = dst_tail.kind()
+                                    && dst_tty.principal().is_some()
```

Comment:
The PR description says the "dst required to have a principal" check will be removed in a future PR, but here it already allows dst not to have a principal and then just doesn't do any checks, not even if auto traits work out. So basically `dyn Trait + 'erased` -> `dyn Send + 'erased` works?

Reply:
"dst required to have a principal" check is in hir_typeck, https://github.com/rust-lang/rust/blob/4b8dbb462c63b9b9235945c0f49b7decf39f0934/compiler/rustc_hir_typeck/src/cast.rs#L918-L920